packages/grafana-ui/src/components/uPlot/utils.ts

```diff
@@ -41,115 +40,219 @@ interface StackMeta {
+      bands.push({
+        series: [si, nextIdx],
+        // fill direction is inverted from stack direction
+        dir: (-1 * dir) as 1 | -1,
+      });
     }
+  });
 
-    collectStackingGroups(f, stackingGroups, seriesIndex);
-    const customConfig: GraphFieldConfig = f.config.custom || {};
+  return bands;
+}
 
-    const values = f.values.toArray();
+// expects an AlignedFrame
+/** @internal */
+export function getStackingGroups(frame: DataFrame) {
+  let groups: Map<string, StackingGroup> = new Map();
 
-    if (customConfig.transform === GraphTransform.NegativeY) {
-      result.push(values.map((v) => (v == null ? v : v * -1)));
-    } else if (customConfig.transform === GraphTransform.Constant) {
-      result.push(new Array(values.length).fill(values[0]));
-    } else {
-      result.push(values);
+  frame.fields.forEach(({ config, values }, i) => {
+    // skip x or time field
+    if (i === 0) {
+      return;
+    }
+
+    let { custom } = config;
+
+    if (custom == null) {
+      return;
+    }
+
+    // TODO: currently all AlignedFrame fields end up in uplot series & data, even custom.hideFrom?.viz
+    // ideally hideFrom.viz fields would be excluded so we can remove this
+    if (custom.hideFrom?.viz) {
+      return;
+    }
+
+    let { stacking } = custom;
+
+    if (stacking == null) {
+      return;
+    }
+
+    let { mode: stackingMode, group: stackingGroup } = stacking;
+
+    // not stacking
+    if (stackingMode === StackingMode.None) {
+      return;
+    }
+
+    // will this be stacked up or down after any transforms applied
+    let vals = values.toArray();
+    let transform = custom.transform;
+    let stackDir =
+      transform === GraphTransform.Constant
+        ? vals[0] > 0
+          ? 1
+          : -1
+        : transform === GraphTransform.NegativeY
+        ? vals.some((v) => v > 0)
+          ? -1
+          : 1
+        : vals.some((v) => v > 0)
+        ? 1
+        : -1;
+
+    let drawStyle = custom.drawStyle;
+    let drawStyle2 =
+      drawStyle === GraphDrawStyle.Bars
+        ? custom.barAlignment
+        : drawStyle === GraphDrawStyle.Line
+        ? custom.lineInterpolation
+        : '';
+
+    let stackKey = [stackDir, stackingMode, stackingGroup, buildScaleKey(config), drawStyle, drawStyle2].join('|');
+
+    let group = groups.get(stackKey);
+
+    if (group == null) {
+      group = {
+        series: [],
+        dir: stackDir,
+      };
+
+      groups.set(stackKey, group);
+    }
+
+    group.series.push(i);
+  });
+
+  return [...groups.values()];
+}
+
+/** @internal */
+export function preparePlotData2(
+  frame: DataFrame,
+  stackingGroups: StackingGroup[],
+  onStackMeta?: (meta: StackMeta) => void
+) {
+  let data = Array(frame.fields.length) as AlignedData;
+
+  let dataLen = frame.length;
+  let zeroArr = Array(dataLen).fill(0);
+  let accums = Array.from({ length: stackingGroups.length }, () => zeroArr.slice());
+
+  frame.fields.forEach((field, i) => {
```

Comment:
the `for` vs `forEach` perf difference only becomes significant for thousands+ of iterations. it would just be noise for anything below 100 and only during setData(), rather than mousemove or resize. if we have 100 fields to deal with, this loop will be the least of our problems. :rofl: